File: javatest/src/main/java/com/example/proxy/java/MyInvocationHandler.java
package com.example.proxy.java;

import java.lang.reflect.InvocationHandler;
import java.lang.reflect.Method;
import java.lang.reflect.Proxy;

/**
 * Created by jaylim on 2017/7/24.
 */

public class MyInvocationHandler implements InvocationHandler {

    private Object target;

    public Object bind(Object tar) {
        this.target = tar;
        return Proxy.newProxyInstance(target.getClass().getClassLoader(), target.getClass().getInterfaces(), this);
    }

    @Override
    public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
        System.out.println(method + " is prepared excute!");
        Object invoke = method.invoke(target, args);
        System.out.println(method + " is already excuted!");
        return invoke;
    }

}

File: javatest/src/main/java/com/example/proxy/MyInvocationHandler.kt
package com.example.proxy

import java.lang.reflect.InvocationHandler
import java.lang.reflect.Method
import java.lang.reflect.Proxy

@Suppress("UNREACHABLE_CODE")
/**
 *
 * Created by jaylim on 2017/7/24.
 */
internal class MyInvocationHandler : InvocationHandler {


    private lateinit var target: Any

    fun bind(tar: Any): Any {
        this.target = tar
        return Proxy.newProxyInstance(target.javaClass.classLoader, target.javaClass.interfaces, this)
    }

    override fun invoke(p0: Any?, p1: Method?, p2: Array<out Any>?): Any {
        println("$p1 is prepared excute!")
        val invoke = p1!!.invoke(target, p2)
        println("$p1 is already excuted!")
        return invoke
    }
}